File: src/Crawler/Analysis/Result/SecurityCheckedHeader.php
<?php

/*
 * This file is part of the SiteOne Crawler.
 *
 * (c) Ján Regeš <jan.reges@siteone.cz>
 */

declare(strict_types=1);

namespace Crawler\Analysis\Result;

class SecurityCheckedHeader
{

    const OK = 1;
    const NOTICE = 2;
    const WARNING = 3;
    const CRITICAL = 4;

    public readonly string $header;
    public ?int $highestSeverity = null;

    /**
     * @var int[] [severity => count]
     */
    public array $countPerSeverity = [];

    /**
     * All unique values of this header
     * @var string[]
     */
    public array $values = [];

    /**
     * @var string[]
     */
    public array $recommendations = [];

    /**
     * @param string $header
     */
    public function __construct(string $header)
    {
        $this->header = $header;
    }

    /**
     * @param array|string|null $value
     * @param int $severity
     * @param string|null $recommendation
     * @return void
     */
    public function setFinding(array|string|null $value, int $severity, ?string $recommendation): void
    {
        if ($value !== null) {
            if (is_array($value)) {
                $value = implode(', ', $value);
            }
            $this->values[$value] = $value;
        }
        if ($recommendation !== null) {
            $this->recommendations[$recommendation] = $recommendation;
        }
        if ($severity > $this->highestSeverity || $this->highestSeverity === null) {
            $this->highestSeverity = $severity;
        }

        if (!isset($this->countPerSeverity[$severity])) {
            $this->countPerSeverity[$severity] = 1;
        } else {
            $this->countPerSeverity[$severity]++;
        }
    }

    public function getFormattedHeader(): string
    {
        $words = explode('-', $this->header);
        foreach ($words as &$word) {
            $word = ucfirst($word);
        }
        return str_replace('Xss', 'XSS', implode('-', $words));
    }

    public function getSeverityName(): string
    {
        return match ($this->highestSeverity) {
            self::OK => 'ok',
            self::NOTICE => 'notice',
            self::WARNING => 'warning',
            self::CRITICAL => 'critical',
            default => 'unknown',
        };
    }

}

File: src/analysis/result/security_checked_header.rs
// SiteOne Crawler - SecurityCheckedHeader
// (c) Jan Reges <jan.reges@siteone.cz>

use std::collections::HashMap;

pub const SEVERITY_OK: i32 = 1;
pub const SEVERITY_NOTICE: i32 = 2;
pub const SEVERITY_WARNING: i32 = 3;
pub const SEVERITY_CRITICAL: i32 = 4;

#[derive(Debug, Clone)]
pub struct SecurityCheckedHeader {
    pub header: String,
    pub highest_severity: Option<i32>,
    /// severity -> count
    pub count_per_severity: HashMap<i32, usize>,
    /// All unique values of this header
    pub values: Vec<String>,
    pub recommendations: Vec<String>,
}

impl SecurityCheckedHeader {
    pub fn new(header: String) -> Self {
        Self {
            header,
            highest_severity: None,
            count_per_severity: HashMap::new(),
            values: Vec::new(),
            recommendations: Vec::new(),
        }
    }

    pub fn set_finding(&mut self, value: Option<&str>, severity: i32, recommendation: Option<&str>) {
        if let Some(val) = value
            && !self.values.contains(&val.to_string())
        {
            self.values.push(val.to_string());
        }
        if let Some(rec) = recommendation
            && !self.recommendations.contains(&rec.to_string())
        {
            self.recommendations.push(rec.to_string());
        }
        if self.highest_severity.is_none() || severity > self.highest_severity.unwrap_or(0) {
            self.highest_severity = Some(severity);
        }
        *self.count_per_severity.entry(severity).or_insert(0) += 1;
    }

    pub fn get_formatted_header(&self) -> String {
        let words: Vec<String> = self
            .header
            .split('-')
            .map(|w| {
                let mut chars = w.chars();
                match chars.next() {
                    Some(c) => format!("{}{}", c.to_uppercase(), chars.as_str()),
                    None => String::new(),
                }
            })
            .collect();
        words.join("-").replace("Xss", "XSS")
    }

    pub fn get_severity_name(&self) -> &'static str {
        match self.highest_severity {
            Some(SEVERITY_OK) => "ok",
            Some(SEVERITY_NOTICE) => "notice",
            Some(SEVERITY_WARNING) => "warning",
            Some(SEVERITY_CRITICAL) => "critical",
            _ => "unknown",
        }
    }
}
